### PR TITLE
refactor(Core/Entities): Cleanup Unit header file

### PR DESCRIPTION
**Pull request by Grimdhex**
## Changes Proposed:

This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

*move several definition in a new file : UnitDefines
*move few class/struct helpers in a new file : UnitUtils
*move code related to CharmInfo in this own file (original idea of Ovahlord, so I credited it even if I didn't cherry-pick this commit)

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Just compile and maybe play a bit

## Known Issues and TODO List:

- [ ] Decide the fate of GlobalCooldownMgr class (look the review comment)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Item/Item.h [hunk 1]
```diff
@@ -35,7 +35,12 @@ struct ItemSetEffect
     SpellInfo const* spells[8];
 };
 
-// EnumUtils: DESCRIBE THIS
```
Comment (Kitzunu) [inline src/server/game/Entities/Item/Item.h:38]: add this back

File: src/server/game/Entities/Unit/CharmInfo.cpp [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+
+    if (_unit->GetTypeId() == TYPEID_PLAYER)                // charmed players don't have spells
+    {
+        //InitEmptyActionBar();
+        return;
+    }
+
+    //InitPetActionBar();
+
+    for (uint32 x = 0; x < MAX_SPELL_CHARM; ++x)
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/CharmInfo.cpp:127]: this is trivial af, but I would consider using `i` instead of `x` for consistency

Comment (Kitzunu) [reply src/server/game/Entities/Unit/CharmInfo.cpp:127]: I know it is a copy paste, but I noticed it now :D

**Comment by Grimdhex** [comment]
@Kitzunu I've some build errors with an output that doesn't seems related to this PR but to your recents changes with ChatHandling (even if I don't see why it's failed now). Have you an idea ?

[ 25%] Building CXX object src/server/scripts/CMakeFiles/scripts.dir/Commands/cs_misc.cpp.o
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/scripts/Commands/cs_message.cpp:25:
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Chat/Channels/Channel.h:23:
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Server/WorldSession.h:30:
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/database/Database/DatabaseEnv.h:21:
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/database/Database/DatabaseWorkerPool.h:23:
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/common/Utilities/StringFormat.h:22:
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/deps/fmt/include/fmt/format.h:49:
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/deps/fmt/include/fmt/core.h:1576:63: fatal error: implicit instantiation of undefined template 'fmt::detail::type_is_unformattable_for<std::optional<std::basic_string<char>>, char>'
    type_is_unformattable_for<T, typename Context::char_type> _;
                                                              ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/deps/fmt/include/fmt/core.h:1808:23: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::basic_format_context<fmt::appender, char>, std::optional<std::basic_string<char>>, 0>' requested here
        data_{detail::make_arg<is_packed, Context>(args)...} {
                      ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/deps/fmt/include/fmt/core.h:1826:10: note: in instantiation of function template specialization 'fmt::format_arg_store<fmt::basic_format_context<fmt::appender, char>, std::optional<std::basic_string<char>>>::format_arg_store<std::optional<std::basic_string<char>>>' requested here
  return {args...};
         ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/deps/fmt/include/fmt/core.h:2788:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::basic_format_context<fmt::appender, char>, std::optional<std::basic_string<char>>>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/common/Utilities/StringFormat.h:52:25: note: in instantiation of function template specialization 'fmt::format<std::optional<std::basic_string<char>> &>' requested here
            return fmt::format(fmt, std::forward<Args>(args)...);
                        ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Chat/Chat.h:158:23: note: in instantiation of function template specialization 'Acore::StringFormatFmt<std::optional<std::basic_string<char>> &>' requested here
        return Acore::StringFormatFmt(GetAcoreString(entry), std::forward<Args>(args)...);
                      ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Chat/Chat.h:152:28: note: in instantiation of function template specialization 'ChatHandler::PGetParseString<std::optional<std::basic_string<char>> &>' requested here
            SendSysMessage(PGetParseString(entry, std::forward<Args>(args)...));
                           ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/scripts/Commands/cs_message.cpp:173:30: note: in instantiation of function template specialization 'ChatHandler::PSendSysMessage<std::optional<std::basic_string<char>> &>' requested here
                    handler->PSendSysMessage(LANG_COMMAND_WHISPEROFFPLAYER, playerNameArg);
                             ^
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/deps/fmt/include/fmt/core.h:1554:45: note: template is declared here
template <typename T, typename Char> struct type_is_unformattable_for;
                                            ^
1 error generated.
gmake[2]: *** [src/server/scripts/CMakeFiles/scripts.dir/build.make:496: src/server/scripts/CMakeFiles/scripts.dir/Commands/cs_message.cpp.o] Error 1

**Comment by Grimdhex** [comment]
This PR is complete now and can be reviewed. I've cleaned up the unnecessary "include" added by my attempts to resolve build errors that were ultimately unrelated to this PR.

**Review by Nyeriah — APPROVED** [review]

**Comment by Nyeriah** [comment]
This PR has introduced the following crash: https://gist.github.com/Nyeriah/73aef2d205cc529795bc2ff319e30e5d

**Comment by Grimdhex** [comment]
> This PR has introduced the following crash: https://gist.github.com/Nyeriah/73aef2d205cc529795bc2ff319e30e5d

Shit, it's related to my pointer comment. I will look this now (I will look if we can move GlobalCooldown without risk in CharmInfo to avoid the pointer change otherwise I revert all the CharmInfo part)

**Pull request merged**